server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java

```diff
@@ -1848,6 +1848,10 @@ public Metadata fromXContent(XContentParser parser) throws IOException {
      * Attempt to find a project for the supplied {@link Index}.
      */
     public Optional<ProjectMetadata> lookupProject(Index index) {
+        if (isSingleProject()) {
+            final var project = projectMetadata.get(DEFAULT_PROJECT_ID);
+            return project.hasIndex(index) ? Optional.of(project) : Optional.empty();
```

Comment:
Technically the optimization can be applied to any single project, not only just the default. But I didn't do it for two reasons:
1. Existing code only handles single-default project
2. No great way to get the single project if it is not the default. We can use `projectMetadata.values().iterator().next()`. But that feels quite a bit overhead since we are talking about a few cpu cycles here.